diff -rupN zeromq-3.2.0/include/zmq.h zeromq-umundo-3.2.0/include/zmq.h
--- zeromq-3.2.0/include/zmq.h	2012-06-05 09:44:53.000000000 +0200
+++ zeromq-umundo-3.2.0/include/zmq.h	2012-07-27 01:50:54.000000000 +0200
@@ -37,6 +37,7 @@ extern "C" {
 #endif
 
 /*  Handle DSO symbol visibility                                             */
+#ifndef ZMQ_STATIC
 #if defined _WIN32
 #   if defined DLL_EXPORT
 #       define ZMQ_EXPORT __declspec(dllexport)
@@ -52,6 +53,9 @@ extern "C" {
 #       define ZMQ_EXPORT
 #   endif
 #endif
+#else
+#define ZMQ_EXPORT
+#endif
 
 /******************************************************************************/
 /*  0MQ versioning support.                                                   */
diff -rupN zeromq-3.2.0/include/zmq_utils.h zeromq-umundo-3.2.0/include/zmq_utils.h
--- zeromq-3.2.0/include/zmq_utils.h	2012-06-05 09:39:30.000000000 +0200
+++ zeromq-umundo-3.2.0/include/zmq_utils.h	2012-07-27 01:50:54.000000000 +0200
@@ -26,6 +26,7 @@ extern "C" {
 #endif
 
 /*  Handle DSO symbol visibility                                             */
+#ifndef ZMQ_STATIC
 #if defined _WIN32
 #   if defined DLL_EXPORT
 #       define ZMQ_EXPORT __declspec(dllexport)
@@ -41,6 +42,9 @@ extern "C" {
 #       define ZMQ_EXPORT
 #   endif
 #endif
+#else
+#define ZMQ_EXPORT
+#endif
 
 /*  Helper functions are used by perf tests so that they don't have to care   */
 /*  about minutiae of time-related functions on different OS platforms.       */
